docs/architecture/adr-033-protobuf-inter-module-comm.md

```diff
@@ -0,0 +1,339 @@
+# ADR 033: Protobuf-based Inter-Module Communication
+
+## Changelog
+
+- 2020-10-05: Initial Draft
+
+## Status
+
+Proposed
+
+## Abstract
+
+This ADR introduces a system for permissioned inter-module communication leveraging the protobuf `Query` and `Msg`
+service definitions defined in [ADR 021](./adr-021-protobuf-query-encoding.md) and
+[ADR 031](./adr-031-msg-service.md) which provides:
+- stable module interfaces to eventually replace the keeper paradigm based on protobuf
+- stronger inter-module object capabilities guarantees
+- module accounts and sub-account authorization
+
+## Context
+
+In the current Cosmos SDK documentation on the [Object-Capability Model](../docs/core/ocap.md), it is state that:
+
+> We assume that a thriving ecosystem of Cosmos-SDK modules that are easy to compose into a blockchain application will contain faulty or malicious modules.
+
+There is currently not a thriving ecosystem of Cosmos SDK modules. We hypothesize that this is in part due to:
+1. lack of a stable v1.0 Cosmos SDK to build modules off of. Module interfaces are changing, sometimes dramatically, from
```

Comment:
I agree, the lack of a stable v1.0 is becoming an increasingly bigger problem. Though it's unclear to me how the design outlined here could help migrate the Cosmos SDK to achieve stability? Isn't this just another way to workaround such longstanding issue?

Reply:
Well, for one buf break check could help. But sure you could have some other way of doing that with plain go. I think the big thing that it helps with is enforcing proper isolation. There are a lot of methods in say the `x/auth` and `x/bank` `Keeper` interfaces that really shouldn't be exposed that generally. So a lot of the benefit here I see is around enforcing proper authorization or "Ocaps"

Reply:
> But sure you could have some other way of doing that with plain go.

Yeah, correct ! That's my point.

> I think the big thing that it helps with is enforcing proper isolation.

But you can have that by strinking the right balance between public packages and `internal/`, can't you?

Reply:
> But you can have that by strinking the right balance between public packages and internal/, can't you?

I mean, in theory that would have been done with existing keepers, but it wasn't and people would be enforcing backwards compatibility with plain go, but we're not. Sometimes, the guard rails need to be built into the protocol. This is a roadmap for that.